File: problems/801_minimum_swaps_to_make_sequences_increasing/solution.py
from typing import List


class Solution:
    def minSwap(self, nums1: List[int], nums2: List[int]) -> int:
        n = len(nums1)
        not_swap = [float('inf')] * n
        swap = [float('inf')] * n
        not_swap[0] = 0
        swap[0] = 1
        for i in range(1, n):
            if nums1[i] > nums1[i - 1] and nums2[i] > nums2[i - 1]:
                not_swap[i] = not_swap[i - 1]
                swap[i] = swap[i - 1] + 1
            if nums1[i] > nums2[i - 1] and nums2[i] > nums1[i - 1]:
                not_swap[i] = min(not_swap[i], swap[i - 1])
                swap[i] = min(swap[i], not_swap[i - 1] + 1)
        return min(not_swap[-1], swap[-1])

File: problems/801_minimum_swaps_to_make_sequences_increasing/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int minSwap(vector<int>& nums1, vector<int>& nums2) {
        int n = nums1.size();
        int notSwap = 0, swap = 1;
        for (int i = 1; i < n; i++) {
            int newNotSwap = INT_MAX, newSwap = INT_MAX;
            if (nums1[i] > nums1[i - 1] && nums2[i] > nums2[i - 1]) {
                newNotSwap = notSwap;
                newSwap = swap + 1;
            }
            if (nums1[i] > nums2[i - 1] && nums2[i] > nums1[i - 1]) {
                newNotSwap = min(newNotSwap, swap);
                newSwap = min(newSwap, notSwap + 1);
            }
            notSwap = newNotSwap;
            swap = newSwap;
        }
        return min(notSwap, swap);
    }
};
